### PR TITLE
Added functionality to monitor network layers activations and weights

### PR DESCRIPTION
**Pull request by SergeyAnufriev**
**Motivation**:

1) Unclear training dynamics, only gradients recorded with losses, while update scheme in the optimizer (equation) involves other variables.  

2) In addition the hypothesis stated that the network might underfit. 

**Hypothesis**:

1) Overfit the network on a small bunch of data. 

    a) In case it fails to achieve close to zero loss, that means the architecture does not have the capacity to learn QM9 dataset.

2) Add network layers activation and weight recordings to monitor:

    a) how many of neurons alive/dead 
    b) weights to gradient ratio (useful for learning rates)
    c) weights magnitude could lead to understanding the future network regularization
    d) layers activation, in the beginning, can indicate if weight initialization was appropriate

The experiment does not require long iterations, so the current problem with slow training will not affect the experiment. 

**Investigation**:

**Experiment 1 - Overfit experiment**

(1) 5 batches of data each with 64 data points were sampled. Reward network was trained from scratch on each batch 100 times. After each network configuration finished training on each dataset, R2 was measured between rewards in batch and predicted. In Fig.1 - h1,h2 are the number of hidden layers in the consequent graph convolution layers, where datasets {0,1,2,3,4} show different batches of data

Experiment code: [link1](https://github.com/SergeyAnufriev/Mol_gan/blob/reward_network/ML_TEST/Overfit_experiment.py)
Configuration parameters file: [link2](https://github.com/SergeyAnufriev/Mol_gan/blob/reward_network/ML_TEST/overfit.yaml)

Fig2b
[Fig.1]

Nerveless network configurations all displayed decreasing the training loss (Fig.2)
Fig2c
[Fig.2]

Fig.3 shows an example of the network-batch configuration with the highest R2. It seems that the network output does not depend on input. 

Fig2a
[Fig.3] 
  
**Observation 1**
    1: In all batches-datasets configurations the network fails to overfit (Fig.,1)
    2: The trained network output does not depend on the input (const predicted reward) (Fig.3)
    3: The network still decreases training loss (Fig.2)

**2. Experiment 2 - Exploring layers activations with gradients weights data**

link3 shows an example of a typical reward network training:

[link3](https://wandb.ai/zcemg08/ml_test/reports/Experiment-2-result--VmlldzozNjI4MDE?accessToken=qssbur3hzckzcfuuml14o4mc9kfe26qp96a7u7brodfcxfbdwte9m3zx336x68di)

**Observation 2**
   1:  The network quickly converges in terms of train and test losses
   2:  Gradients in each layer subsequent by backpropagation order decrease in magnitude (When initialised)
   3:  Activations of each layer ordered by the forward pass decrease variance in output values 
   4: Last layer activation has distribution with zero mean, hence the sigmoid will remove half of the activations of the neurons. 

**Experiment 3 - 20 epoch training**

The experiment had grid search over the number of hidden layers in convolutional layers [h1,h2] and different learning rates. The network was trained for 20 epochs and test loss with R2 for predicted values on the whole test set were recorded [Table.1]

bz  |  h1   |  h2   |  h3   |  h4  |  epochs  |  drop_out  |  learning_rate  |  Train_loss   |  Test_Loss    |  R2_TEST
----|-------|-------|-------|------|----------|------------|-----------------|---------------|---------------|-------------
64  |  256  |  64   |  128  |  54  |  20      |  0.1       |  0.002          |  0.000466048  |  0.000593351  |  0.737936855
64  |  128  |  64   |  128  |  54  |  20      |  0.1       |  0.002          |  0.000980526  |  0.000606668  |  0.732065853
64  |  128  |  64   |  128  |  64  |  20      |  0.1       |  0.001          |  0.000285601  |  0.000612252  |  0.729258461
64  |  128  |  128  |  128  |  54  |  20      |  0.1       |  0.002          |  0.000557845  |  0.000627796  |  0.722741199
64  |  128  |  64   |  128  |  54  |  20      |  0.1       |  0.005          |  0.001025601  |  0.000675039  |  0.701870101
64  |  128  |  64   |  128  |  64  |  20      |  0.1       |  0.0005         |  0.000397991  |  0.000695878  |  0.692278689
64  |  256  |  64   |  128  |  54  |  20      |  0.1       |  0.005          |  0.000675906  |  0.000747617  |  0.669802645
64  |  256  |  128  |  128  |  54  |  20      |  0.1       |  0.005          |  0.000631543  |  0.000941729  |  0.5840852
64  |  128  |  128  |  128  |  54  |  20      |  0.1       |  0.005          |  0.001096054  |  0.001005034  |  0.556138615

[Table.1]

Parallel coordinates plot Fig.5 was created for Table.1 

W B Chart 12_16_2020, 11_48_57 AM
 [Fig.5]

However compared to experiment 2, where the network was overfitted on each batch of data. The network did not seem to show output being independent of the input. (Fig. 7)

Best_so_far
[Fig.7]

**Observation***

The network seems to be sensible on learning rates values (Fig.7) with the optimum LR values in the range between 1*e-3 to 2*e-3. While the best 3 results in terms of R2 tets and test loss have the same network configuration in terms of the number of convolution hidden layers. 

**Conclusion**

The next experiment must use network configuration as in the best 2 results (Table) with different learning rate schedules. 
The baseline for R2 on the whole test so far 0.7379 (Table.1)

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -180,19 +180,23 @@ def __init__(self,in_channels,h_1,h_2,h_3,h_4,drop_out,device):
     self.lin4   = nn.Linear(h_3,h_4,bias=True)
     self.lin5   = nn.Linear(h_4,1,bias=True)
 
-    self.act        = nn.Tanh()
-    self.act_last   = nn.Sigmoid()
+    self.act1        = nn.Tanh()
+    self.act2        = nn.Tanh()
+    self.act3        = nn.Tanh()
+    self.act4        = nn.Tanh()
+    self.act5        = nn.Tanh()
```
Comment (markus-hinsche) [inline model.py:187]: You could use the same activation object instead of creating many. This is similar to here:
https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L72-L81

File: Diagnostic.py [hunk 2]
```diff
@@ -1,332 +1,352 @@
...
+  
+def min_max(X): #X 2D array
+  x1_gen = X[:,0]
+  x2_gen = X[:,1]
+  return np.min(x1_gen),np.max(x2_gen),\
+         np.min(x2_gen),np.max(x2_gen)
+
+'''vis function plots GAN 2D generated vs real points, with background coloured by discriminator output'''
+
+def vis(G,D,X_train_save,z_fixed,device):
```
Comment (markus-hinsche) [inline Diagnostic.py:307]: It is had to understand what `G` and `D` mean. Maybe you can better names for your variables.
Also local variables and parameters are by convention lowercase

File: Diagnostic.py [hunk 3]
```diff
@@ -1,332 +1,352 @@
...
+  def rand_vec(self):
+    v = torch.cat([torch.randint_like(p.data, high=2,device=self.device).flatten()\
+          for p in self.params]).unsqueeze(-1)
+    v[v==0.]=-1. 
+    return v/np.sqrt(len(v))
+
+  '''Orthogonilise: Gram-Schmidt process '''
+
+  @staticmethod
+  def ort(w,list1):
```
Comment (markus-hinsche) [inline Diagnostic.py:249]: Put docstrings into their respective function to increase readability
```
  @staticmethod
  def ort(w,list1):
    '''Orthogonilise: Gram-Schmidt process '''
```

**Review by markus-hinsche — COMMENTED** [review]
A few suggestion how to make the code more readable for your colleagues :)

File: model.py (hunk 1 above)
Comment (SergeyAnufriev) [reply model.py:187]: I made them separate, so I can track activations from different layers. I will have a look again how to use PyTorch hooks

File: Diagnostic.py (hunk 3 above)
Comment (SergeyAnufriev) [reply Diagnostic.py:249]: Notations fixed
